File: Programmers/python/완주하지 못한 선수.py
# 1번 풀이
def solution(participant, completion):
    answer = ''
    participant_dict = {}
    
    for person in participant:
        if person not in participant_dict:
            participant_dict[person] = 1
        else:
            count = participant_dict[person]
            participant_dict.update({person: count + 1})
    
    for person in completion:
        count = participant_dict[person]
        if count == 1:
            del participant_dict[person]
        else:
            participant_dict.update({person: count - 1})
    
    lists = list(participant_dict.keys())
    
    return lists[0]


# 2번 풀이 - sort(), zip() 활용
def solution(participant, completion):   
    participant.sort()
    completion.sort()
    
    for p, c in zip(participant, completion):
        if p != c:
            return p
        
    return participant[-1]

  
# 3번 풀이 - Counter 객체 활용
from collections import Counter

def solution(participant, completion):   
    out = Counter(participant) - Counter(completion)
    return list(out)[-1]

File: Programmers/java/완주하지 못한 선수.java
import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        String answer = "";
        Map<String, Integer> people = new HashMap<>();
        
        for(String person : participant) {
            if(!people.containsKey(person)) {
                people.put(person, 1);
                continue;
            }
            int cur = people.get(person);
            people.put(person, cur + 1);
        }
        
        for(String archive : completion) {
            int cur = people.get(archive);
            people.put(archive, cur - 1);
        }
        
        for(Map.Entry<String, Integer> entry : people.entrySet()) {
            if(entry.getValue() != 0) {
                return entry.getKey();
            }
        }
        
        return answer;
    }
}
